### DISCUSSION TITLE
Any plan to support the Executable Directive Locations ?

### DISCUSSION THREAD

**Discussion by Yao544303** [graphql-java/graphql-java#3311] [Q&A]
Currently, seems it only supports the [TypeSystemDirectiveLocation](https://www.graphql-java.com/documentation/sdl-directives),
is there any plan to support the [ExecutableDirectiveLocation](https://spec.graphql.org/October2021/#ExecutableDirectiveLocation) in the future? 

Or is any idea to implement ExecutableDirectiveLocation?
For example, here is the schema.graphqls
`
directive @uppercase on FIELD
type Query {
    bookById(id: ID): Book
}

type Book {
    id: ID
    name: String
    pageCount: Int
    author: String
}`

Need return 
`{
  "data": {
    "bookById": {
      "id": "book-1",
      "name": "Effective Java",
      "pageCount": 416,
      "author": "author-1",
    }
  }
}`

for query
`query bookDetails {
  bookById(id: "book-1") {
    id
    name
    pageCount
    author 
  }
}`

And return
`{
  "data": {
    "bookById": {
      "id": "book-1",
      "name": "Effective Java",
      "pageCount": 416,
      "author": "AUTHOR-1",
    }
  }
}`

for query
`query bookDetails {
  bookById(id: "book-1") {
    id
    name
    pageCount
    author @uppercase 
  }
}`

Thanks.

**Comment by dondonz** [comment]
Hi @Yao544303, graphql-java supports both TypeSystemDirectiveLocation (for Schema Directives) and ExecutableDirectiveLocation (for Query Directives, also known as Operation Directives). You can see all the supported directive locations in this enum: https://github.com/graphql-java/graphql-java/blob/edd26ccb957b65c6669782961830b5d74abfd43e/src/main/java/graphql/introspection/Introspection.java#L481

Thanks for pointing out that our documentation only seems to refer to Schema Directives! I will add a new page for Query/Operation Directives as well. I like the example you provided, I'll build on that.

It will take me a while to write a page with a tutorial, but in the meantime if anybody has an example, feel free to add to this discussion.

**Comment by jhodge-chwy** [comment]
+1 would also like the above

**Comment by piotrblasiak** [comment]
Any update on this? I can't find any example on how to implement Query directives.

**Reply by bbakerman** [reply]
So nominally when we say graphql-java supported Operation directives... it means we can capture them and tell you about them being present on an element.

However the challenge is the execution aspects of such a directive.  Imagine you wanted a `@cached` directive that told the server to load from cache if it can.  Well graphql-java can tell you about them existing via [QueryDirectives](https://github.com/graphql-java/graphql-java/blob/master/src/main/java/graphql/execution/directives/QueryDirectives.java) class but its up to you to decide how to implement its execution meaning.  eg in your data fetchers you might decide to use some. caching for data say - graphql-java cant do that - but it can tell you an operation (aka query) directive is present

**Reply by bbakerman** [reply]
`DataFetchingEnvironment` has a `QueryDirectives` property that shows you the directives that are on on the current field

**Reply by bbakerman** [reply]
The question back at you is - what do you achieve with a query directive... I can maybe help you and  point you in a direction

**Reply by piotrblasiak** [reply]
Ok that makes sense - it works for me except for in one situation:

I'm implementing an @await directive, that will wait for some conditions - for example for a payment to get confirmed by a stripe webhook etc.
The datafetcher approach should work, except if I would like to put the directive directly on the Query:

```
query MyQuery($awaitConditions: AwaitConditionsInput) @await(conditions: $awaitConditions) {
// ... now after the await finished, then start fetching all fields.
}
```

It is a very special situation, and I have been able to solve it by extending the QueryExecutionStrategy but then I had to implement the Value resolution etc myself.

The normal case of:

```
query MyQuery($id: ID!) {
  payment(id: $id, confirmed: true) @await {
    someOtherField
  }
}
```

Will work with the datafetcher approach.

**Comment by xghost404** [comment]
yeahh

**Comment by dondonz** [comment]
Query/operation directive documentation is coming soon! https://github.com/graphql-java/graphql-java-page/pull/185
